### PR TITLE
Pull schema templates from ACR and generate Parquet schema

### PR DESCRIPTION
**Pull request by QuanWanxx**
1. Add ability to pull schema templates from Azure Container Registry.
2. Add ability to generate Parquet schema from required valid Json schema.
    1) All properties in Json schema should be basic types.

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/AzureContainerRegistryAccessTokenProvider.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+namespace Microsoft.Health.Fhir.Synapse.SchemaManagement.ContainerRegistry
+{
+    /// <summary>
+    /// Retrieve ACR access token with AAD token provider.
+    /// We need to exchange ACR refresh token with AAD token, and get ACR access token from refresh token.
+    /// References:
+    /// https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#calling-post-oauth2exchange-to-get-an-acr-refresh-token
+    /// https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#calling-post-oauth2token-to-get-an-acr-access-token
+    /// </summary>
+    public class AzureContainerRegistryAccessTokenProvider : IContainerRegistryTokenProvider
```
Comment (yankunhuang-pku) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/AzureContainerRegistryAccessTokenProvider.cs:28]: If this class is similar to AzureContainerRegistryAccessTokenProvider in fhir server lib, can we try to leverage that class. If not, can we consider using a different class name?

File: FhirToDataLake/NuGet.Config [hunk 2]
```diff
@@ -4,5 +4,6 @@
     <clear />
     <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
     <add key="FhirAnalyticsPublic" value="https://microsofthealthoss.pkgs.visualstudio.com/FhirAnalytics/_packaging/FhirAnalyticsPublic/nuget/v3/index.json" />
+    <add key="Microsoft Health OSS" value="https://microsofthealthoss.pkgs.visualstudio.com/FhirServer/_packaging/Public/nuget/v3/index.json" />
```
Comment (yankunhuang-pku) [inline FhirToDataLake/NuGet.Config:7]: Can we keep the same format without blank?

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.SchemaManagement.UnitTests/Parquet/FhirParquetSchemaManagerTests.cs [hunk 3]
```diff
@@ -31,7 +33,7 @@ static FhirParquetSchemaManagerTests()
         [Theory]
         public static void GivenASchemaType_WhenGetSchema_CorrectResultShouldBeReturned(string schemaType, int propertyCount)
         {
-            var schemaManager = new FhirParquetSchemaManager(_schemaConfigurationOption, NullLogger<FhirParquetSchemaManager>.Instance);
+            var schemaManager = new FhirParquetSchemaManager(_schemaConfigurationOption, _jsonSchemaCollectionsProvider, NullLogger<FhirParquetSchemaManager>.Instance);
```
Comment (yankunhuang-pku) [inline FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.SchemaManagement.UnitTests/Parquet/FhirParquetSchemaManagerTests.cs:36]: If we would use shcemaManager in every/most tests, can we create it only once before tests?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/JsonSchemaCollectionsProvider.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            /*
+            List<Dictionary<string, Template>> templateCollections = await GetTemplateCollectionsAsync(schemaImageReference, cancellationToken);
+            */
+
+            // Will implement later when integrating the FHIR-Converter.
+            // Will do a try drive among templates for all resource types with empty Json data, to get involved Json schemas.
+            throw new NotImplementedException();
+        }
+
+        private async Task<List<Dictionary<string, Template>>> GetTemplateCollectionsAsync(string schemaImageReference, CancellationToken cancellationToken)
```
Comment (yankunhuang-pku) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/JsonSchemaCollectionsProvider.cs:62]: GetTemplateCollectionsAsync should be in another class if call it as JsonSchemaCollectionsProvider?

Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/JsonSchemaCollectionsProvider.cs:62]: This is a private method to get template collections using the inner **_templateCollectionProviderFactory**.
The ```JsonSchemaCollectionsProvider``` will have only 1 public interface ```GetJsonSchemaCollectionAsync``` return the Json schema collection, but it temporarily not implemented.

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.SchemaManagement.UnitTests/Parquet/FhirParquetSchemaManagerTests.cs (hunk 3 above)
Comment (QuanWanxx) [reply FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.SchemaManagement.UnitTests/Parquet/FhirParquetSchemaManagerTests.cs]: Updated.

File: FhirToDataLake/NuGet.Config (hunk 2 above)
Comment (QuanWanxx) [reply FhirToDataLake/NuGet.Config:7]: The spaces in NuGet source key? I follow up the key in fhir-server project [nuget.config](https://github.com/microsoft/fhir-server/blob/main/nuget.config)

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/AzureContainerRegistryAccessTokenProvider.cs (hunk 1 above)
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/AzureContainerRegistryAccessTokenProvider.cs]: Updated to **ContainerRegistryAccessTokenProvider**

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/JsonSchemaCollectionsProvider.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Synapse.SchemaManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement;
+using Microsoft.Health.Fhir.TemplateManagement.Exceptions;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+using Newtonsoft.Json.Schema;
+
+namespace Microsoft.Health.Fhir.Synapse.SchemaManagement.ContainerRegistry
+{
+    public class JsonSchemaCollectionProvider
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/ContainerRegistry/JsonSchemaCollectionsProvider.cs:23]: I feel this class works more like a implementation of some `ISchemaProvider`, if we are not using the JSchema directly, we can add the interface like `Dictionary<string, FhirSchemaNode> GetSchemaAsync()`, then we can have two different implementations, local schema provider and acr schema provider.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/CustomizedSchema/JsonSchemaParser.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                throw new ParseJsonSchemaException(string.Format("The \"{0}\" customized schema have no \"type\" keyword or \"type\" is null.", resourceType));
+            }
+
+            if (jSchema.Type != JSchemaType.Object)
+            {
+                throw new ParseJsonSchemaException(string.Format("The \"{0}\" customized schema type \"{1}\" should be \"object\".", resourceType, jSchema.Type));
+            }
+
+            string schemaType = $@"{resourceType}_Customized";
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/CustomizedSchema/JsonSchemaParser.cs:26]: do we want to hard code the schema type here or read it from the schema data?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 7]
```diff
@@ -118,16 +146,32 @@ public FhirParquetSchemaNode GetSchema(string schemaType)
...
+        private Dictionary<string, FhirParquetSchemaNode> LoadCustomizedSchemas(string schemaImageReference)
+        {
+            var jSchemaCollections = _jsonSchemaCollectionsProvider.GetJsonSchemaCollectionAsync(schemaImageReference, CancellationToken.None).Result;
+            _logger.LogInformation($"{jSchemaCollections.Count} resource customized schemas been fetched from {schemaImageReference}.");
+
+            var customizedSchemaNodesMap = new Dictionary<string, FhirParquetSchemaNode>();
+
+            foreach (var jSchemaItem in jSchemaCollections)
+            {
+                var parquetSchemaNode = _jsonSchemaParser.ParseJSchema(jSchemaItem.Key, jSchemaItem.Value);
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs:170]: Do we assume the `jSchemaItem.Key` here is always a resource type?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 8]
```diff
@@ -8,10 +8,13 @@
 using System.IO;
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs]: Updated.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 9]
```diff
@@ -20,24 +23,55 @@ public class FhirParquetSchemaManager : IFhirSchemaManager<FhirParquetSchemaNode
...
         public FhirParquetSchemaManager(
             IOptions<SchemaConfiguration> schemaConfiguration,
+            ParquetSchemaProviderDelegate parquetSchemaDelegate,
             ILogger<FhirParquetSchemaManager> logger)
         {
             _logger = logger;
 
-            _resourceSchemaNodesMap = LoadSchemas(schemaConfiguration.Value.SchemaCollectionDirectory);
-            _logger.LogInformation($"Initialize FHIR schemas completed, {_resourceSchemaNodesMap.Count} resource schemas been loaded.");
+            _defaultSchemaProvider = parquetSchemaDelegate("default");
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs]: Sure, updated.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 10]
```diff
@@ -20,24 +23,55 @@ public class FhirParquetSchemaManager : IFhirSchemaManager<FhirParquetSchemaNode
...
+            _customizedSchemaProvider = parquetSchemaDelegate("custom");
+
+            _resourceSchemaNodesMap = _defaultSchemaProvider.GetSchemasAsync(schemaConfiguration.Value.SchemaCollectionDirectory).Result;
+            _logger.LogInformation($"{_resourceSchemaNodesMap.Count} resource default schemas have been loaded.");
+
+            if (_customSchema)
+            {
+                if (string.IsNullOrEmpty(schemaConfiguration.Value.CustomizedSchemaImageReference))
+                {
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs]: Sure, updated.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 11]
```diff
@@ -20,24 +23,55 @@ public class FhirParquetSchemaManager : IFhirSchemaManager<FhirParquetSchemaNode
...
+                    throw new FhirSchemaException("When enable customized schema, schema image reference cannot be null or empty.");
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs]: Have moved to ```ConfigurationErrorException``` and fixed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs [hunk 12]
```diff
@@ -20,24 +23,55 @@ public class FhirParquetSchemaManager : IFhirSchemaManager<FhirParquetSchemaNode
...
-                _schemaTypesMap.Add(resourceType, new List<string>() { resourceType });
+                if (_schemaTypesMap.ContainsKey(schemaNodeItem.Value.Type))
+                {
+                    _schemaTypesMap[schemaNodeItem.Value.Type].Add(schemaNodeItem.Key);
+                }
+                else
+                {
+                    _schemaTypesMap.Add(schemaNodeItem.Value.Type, new List<string>() { schemaNodeItem.Key });
+                }
             }
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/FhirParquetSchemaManager.cs]: Have implemented this in ```ACRCustomizedSchemaProvider```, remove comment here

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/SchemaProvider/ACRCustomizedSchemaProvider.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Synapse.SchemaManagement.ContainerRegistry;
+using Newtonsoft.Json.Schema;
+
+namespace Microsoft.Health.Fhir.Synapse.SchemaManagement.Parquet.SchemaProvider
+{
+    public class ACRCustomizedSchemaProvider : IParquetSchemaProvider
```
Comment (moria97) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.SchemaManagement/Parquet/SchemaProvider/ACRCustomizedSchemaProvider.cs:17]: Mightbe `AcrCustomizedSchemaProvider`?

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/DataProcessor/ParquetDataProcessorTests.cs [hunk 14]
```diff
@@ -38,14 +41,19 @@ static ParquetDataProcessorTests()
             {
                 SchemaCollectionDirectory = TestUtils.DefaultSchemaDirectoryPath,
             });
-
-            _fhirSchemaManager = new FhirParquetSchemaManager(schemaConfigurationOption, NullLogger<FhirParquetSchemaManager>.Instance);
+
```
Comment (moria97) [inline FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/DataProcessor/ParquetDataProcessorTests.cs:44]: extra spaces here

**Review by moria97 — APPROVED** [review]

File: FhirToDataLake/build.yml [hunk 15]
```diff
@@ -77,6 +79,8 @@ steps:
     codeCoverageEnabled: true
     platform: '$(buildPlatform)'
     configuration: '$(buildConfiguration)'
+  env:
+    TestContainerRegistryPassword: $(TestContainerRegistryPasswordSecret)
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/build.yml]: Moved those 2 secret values to KeyVault.

**Pull request merged**